src/main/java/org/jetbrains/research/kotlincodesmelldetector/core/distance/DistanceMatrix.kt

```diff
@@ -1,5 +1,256 @@
 package org.jetbrains.research.kotlincodesmelldetector.core.distance
 
-class DistanceMatrix {
+import com.intellij.openapi.progress.ProgressIndicator
+import org.jetbrains.kotlin.idea.refactoring.fqName.getKotlinFqName
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.collectDescendantsOfType
+import org.jetbrains.kotlin.psi.psiUtil.containingClassOrObject
+import org.jetbrains.kotlin.psi.psiUtil.forEachDescendantOfType
+import org.jetbrains.research.kotlincodesmelldetector.KotlinCodeSmellDetectorBundle
+import org.jetbrains.research.kotlincodesmelldetector.utils.getConstructorType
+import org.jetbrains.research.kotlincodesmelldetector.utils.getGenericType
+import org.jetbrains.research.kotlincodesmelldetector.utils.isContainer
+import org.jetbrains.research.kotlincodesmelldetector.utils.nameWithParameterList
+import java.util.*
+import kotlin.collections.ArrayList
 
-}
+class DistanceMatrix(private val project: ProjectInfo) {
+    val classes: MutableMap<FqName, ClassEntity> = mutableMapOf()
+    private val entityIndexMap = mutableMapOf<FqName, Int>()
+    private val classIndexMap = mutableMapOf<FqName, Int>()
+    private val classList = mutableListOf<ClassEntity>()
+    private val entityList = mutableListOf<KtNamedDeclaration>()
+    private val entityMap = mutableMapOf<FqName, MutableSet<FqName>>()
+    private val classMap = mutableMapOf<FqName, Set<FqName>>()
+    private val maximumNumberOfSourceClassMembersAccessedByMoveMethodCandidate = 2
+
+    init {
+        generateEntitySets()
+        generateDistances()
+    }
+
+    private fun generateEntitySets() {
+        for (ktClassPointer in project.classes) {
+            val ktClass = ktClassPointer.element!!
+            if (ktClass is KtClassOrObject) {
+                if (!(ktClass is KtClass && ktClass.isEnum())) {
+                    val classEntity = ClassEntity(ktClass)
+                    classes[classEntity.fqName] = classEntity
+                }
+            }
+        }
+        for (myClass in classes.values) {
```

Comment:
To many nested levels, hard to understand what's going on. Splitting this loop internals into functions will help